### ISSUE TITLE
Comparing two operations that contain `log` sometimes leads to TypeError exception

### ISSUE THREAD

**Issue by jonatanschroeder** [sympy/sympy#22020]
This originally comes from PrairieLearn/PrairieLearn#4793, but it seems the problem is in sympy. We weren't able to narrow it down much further, other than to identify that this is specific to `log` (other functions like square root don't seem to have this problem).

In essence, two expressions can lead to an exception when being compared if the `log` expression is multiplied by a variable instead of a fixed coefficient.

```
Python 3.8.5 (default, May 27 2021, 13:30:53)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sympy.parsing.sympy_parser import parse_expr
>>> x = parse_expr("log((2*V/3-V)/C)/-(R+r)*C")
>>> y = parse_expr("log((2*V/3-V)/C)/-(R+r)*2")
>>> x.equals(y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/core/expr.py", line 751, in equals
    constant = diff.is_constant(simplify=False, failing_number=True)
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/core/expr.py", line 674, in is_constant
    if b is not None and b is not S.NaN and b.equals(a) is False:
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/core/expr.py", line 741, in equals
    diff = factor_terms(simplify(self - other), radical=True)
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/simplify/simplify.py", line 635, in simplify
    _e = cancel(expr)
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/polys/polytools.py", line 6688, in cancel
    f = factor_terms(f, radical=True)
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/core/exprtools.py", line 1274, in factor_terms
    return do(expr)
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/core/exprtools.py", line 1271, in do
    rv = _keep_coeff(cont, p, clear=clear, sign=sign)
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/core/mul.py", line 2069, in _keep_coeff
    if r == int(r):
  File "/home/jonatan/.local/lib/python3.8/site-packages/sympy/core/expr.py", line 332, in __int__
    raise TypeError("can't convert %s to int" % r)
TypeError: can't convert oo to int
>>>
```

**Comment by oscargus** [comment]
The problem comes from https://github.com/sympy/sympy/blob/5331d056276f22b74587c04cac8399b3ddfc6048/sympy/core/expr.py#L661-L669

where all free symbols are replaced by zeros on line 666 and numerically evaluated,which results in `-oo`. This later breaks the checking. I assume that checking that a is not infinite at line 669 will fix your problem, but not sure if it will break something else...

**Comment by echuber2** [comment]
This doesn't happen in Sympy 1.6.2 but does in 1.8. Is it not a regression then?

**Comment by echuber2** [comment]
To elaborate, in 1.6.2 instead of getting an error, the `equals` call returns `None`:

``` python
>>> import sympy
>>> from sympy.parsing.sympy_parser import parse_expr
>>> x = parse_expr("log((2*V/3-V)/C)/-(R+r)*C")
>>> x
C*log(-V/(3*C))/(-R - r)
>>> y = parse_expr("log((2*V/3-V)/C)/-(R+r)*2")
>>> y
2*log(-V/(3*C))/(-R - r)
>>> x.equals(y)
>>> print(x.equals(y))
None
```

**Comment by oscargus** [comment]
> Is it not a regression then?

I'd say it is, yes. This code has not been touched in the last seven years, so probably something with the numerical evaluation has changed so that oo is returned rather than throwing a ZeroDivisionError.

**Comment by oscargus** [comment]
Added a fix in #22022. Hopefully it doesn't break anything else.

`None` is sort of correct here as sympy cannot really determine if the expressions are the same or not. (I assume that we do not want to return `False` as they can be identical if C is 2, but not really sure about the logic.) And it should definitely not throw an error.

**Comment by oscarbenjamin** [comment]
I bisected this to aef0304b86a5913b0dcf34483998277efff1cb46 from #20885. Maybe that should be changed or reverted.

**Comment by oscargus** [comment]
>  Maybe that should be changed or reverted.

I've realized that it still makes sense to have these checks. If https://github.com/sympy/sympy/blob/c181470266243ffdefa3c4129165e0e1e009195f/sympy/core/expr.py#L662-L663
returns `oo`, I think the same thing will happen. So I wouldn't fully blame #20885 here.

**Comment by oscargus** [comment]
One could of course discuss how `is_constant` should work in general.

**Comment by oscargus** [comment]
The underlying problem is:

```
In [7]: simplify(-log(3)/2 + I*pi/2 -oo)
Traceback (most recent call last):

  File "C:\Users\Oscar\AppData\Local\Temp/ipykernel_14780/1626714943.py", line 1, in <module>
    simplify(-log(3)/2 + I*pi/2 -oo)

  File "C:\Users\Oscar\sympy\sympy\simplify\simplify.py", line 635, in simplify
    _e = cancel(expr)

  File "C:\Users\Oscar\sympy\sympy\polys\polytools.py", line 6682, in cancel
    f = factor_terms(f, radical=True)

  File "C:\Users\Oscar\sympy\sympy\core\exprtools.py", line 1274, in factor_terms
    return do(expr)

  File "C:\Users\Oscar\sympy\sympy\core\exprtools.py", line 1271, in do
    rv = _keep_coeff(cont, p, clear=clear, sign=sign)

  File "C:\Users\Oscar\sympy\sympy\core\mul.py", line 2115, in _keep_coeff
    if any(c == int(c) for c, _ in args):

  File "C:\Users\Oscar\sympy\sympy\core\mul.py", line 2115, in <genexpr>
    if any(c == int(c) for c, _ in args):

  File "C:\Users\Oscar\sympy\sympy\core\expr.py", line 333, in __int__
    raise TypeError("can't convert %s to int" % r)

TypeError: can't convert -oo to int
```

**Comment by oscargus** [comment]
Better fix in #22024
